src/index.js

```diff
@@ -1,58 +1,74 @@
-import React from "react";
-import PropTypes from "prop-types";
-import { useSelector } from "react-redux";
-import { withRouter } from "react-router-dom";
+import React, { useContext } from 'react';
+import PropTypes from 'prop-types';
 
 /**
- * This function instantiates a React.Component with a set of additional passed props. It supports
- * `Overridable` components.
- *
- * @param {React.Component} Component - the actual component that will be parametrized
- * @param {Object|Function} extraProps - additional properties passed to the final component
+ * Object responsible of keeping track of all overridden components
+ * @constructor object containing the initial map `id: Component` of overriden components
+ */
+class OverriddenComponentsStore {
+  constructor(overriddenCmps) {
+    this.components = overriddenCmps || {};
+  }
+
+  add = (id, Component) => {
+    const alreadyOverridden = id in this.components;
+    if (alreadyOverridden) {
+      throw Error(`The component ${id} has been already overridden.`);
+    }
+    this.components[id] = Component;
+  };
+
+  get = (id) => {
+    return this.components[id];
+  };
+
+  clear = () => {
+    this.components = {};
+  };
+}
+export const overrideStore = new OverriddenComponentsStore();
+export const OverridableContext = React.createContext(overrideStore);
```

Comment:
The default passed to `createContext` is only used when no provider for the context is present - so I think we need to decide which behavior we want in this case:

- do not override anything if the context  is not provided
- always apply defined overrides, regardless of whether or not the context is provided

If we go for the first option, I wouldn't pass a default here.

If we go for the second option, then I propose removing the context altogether **or** being clear that the context provider should usually *not* be specified, unless you want to use a custom store for your overrides instead of the global one.

Reply:
Agree with you, I would remove the default.
The main question is about your previous comment and this is exactly what I was trying to explain in the notes.
When I took the `Overridable` from Indico and adapted to our needs, I removed hooks+redux and replaced with a global store everywhere, no context. Personally, I am not very happy with this global store and I find Context nicer, but it solves our problem.

With global store you can do from anywhere, any js file this:
```
import {store} ...
store.add(id, MyCustomCmp)
```
and then in our internal `app.js` we don't have to do anything because `Overridable` will use the global store.

With React Context, you will have to pass the map to the `<Context.Provider value={map}`. Very easy in a SPA, but in Invenio is harder: when building JS assets, the internal `app.js` does not know if the user has created an extra `.js` to override components. Basically I cannot have in `app.js` (like in Python for example if you allow me the pseudocode to explain what I mean):
```
try {
  import overrides from './overrides.js';
} catch {
   // no overrides
}
...
<Context.Provider value={overrides}...
```
I would have to inject this extra obj `overrides` via HTML DOM... Is it maybe more clear? Let me know if you have a better solution in mind...

Reply:
I think you could always have the overrides locally in your application and just import it from both the place where you pass it to the <Context.Provider>` and the places where you want to add new overrides.

But I do see the benefit of having a global list of overrides. ;)

Reply:
But at least when you use plugins there's also an argument against the global override list: If you define one entry point for all plugins and let them "register" their overrides, you really don't want to have to deal with shared state - especially since it may not work well when the plugins and core are built as separate webpack projects.